core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -142,16 +142,17 @@ private[log] class ProducerIdEntry(val producerId: Long, val batchMetadata: muta
  *                                of this is the consumer offsets topic which uses producer ids from incoming
  *                                TxnOffsetCommit, but has no sequence number to validate and does not depend
  *                                on the deduplication which sequence numbers provide.
- * @param loadingFromLog This parameter indicates whether the new append is being loaded directly from the log.
- *                       This is used to repopulate producer state when the broker is initialized. The only
- *                       difference in behavior is that we do not validate the sequence number of the first append
- *                       since we may have lost previous sequence numbers when segments were removed due to log
- *                       retention enforcement.
+ * @param isFromClient The parameter indicates whether the write is coming from a client or not. If it is not coming
+ *                     from a client, it could be due to replication traffic, or when rebuilding producer state on
+ *                     from the log. In the latter two cases, we should not validate the append, but accept the
+ *                     incoming append unconditionally. This is for two reasons: first, the write was already
+ *                     validated when received from the client. Second, the data is already the log, so it is not
+ *                     clear what would be achieved by validating it again.
  */
 private[log] class ProducerAppendInfo(val producerId: Long,
                                       currentEntry: ProducerIdEntry,
                                       validateSequenceNumbers: Boolean,
-                                      loadingFromLog: Boolean) {
+                                      isFromClient: Boolean) {
```

Comment:
There's a bit of overlap between `validateSequenceNumbers` and `isFromClient`. The only difference is that `isFromClient` also skips the fencing check. Would it make sense to make the check `validateEpoch` or something like that for consistency?

Reply:
I agree that this is definitely a bit tricky. The problem is that the flags are sort of hierarchical: 

When the request is from the client, but for the consumer offsets topic, we _only_ want to validate the epoch, but not the sequence numbers. In this case, `validateSequenceNumbers` will be false.

When the request isn't from the client, we don't want to do any validation. In this case, it doesn't matter what the value of `validateSequenceNumbers` is.

Not sure if there is a clearly superior way to represent these kinds of hierarchical flags in code, other than creating an enum type like `ValidationType`. This could have values `NoValidation`, `EpochOnlyValidation`, `FullValidation`. 

Update: I tried to implement this, but it's a bit painful with scala enums :(

Reply:
I added a `ValidationType` pseudo-enum (using sealed traits and case objects per @ijuma's suggestion). This makes the code cleaner IMO.